### PR TITLE
UPG counters

### PR DESCRIPTION
**Pull request by sergeymatov**
This commit adds small set of UPG-related counters.
Based on vlib_counter lib.

### CODE REVIEW

**Comment by ivan4th** [comment]
It should be rather easy to verify the counters from test_upf.py, so probably that should be done

File: upf/flowtable_init.c [hunk 1]
```diff
@@ -80,6 +83,14 @@ flowtable_init_cpu (flowtable_main_t * fm, flowtable_main_per_cpu_t * fmt)
...
   upf_debug ("POOL SIZE %u", pool_elts (fmt->timers));
 
+  /* Init flows counter per cpu */
+  upf_counter_lock (gtm);
+  vlib_validate_simple_counter (&gtm->upf_simple_counters[UPF_FLOW_COUNTER],
+				fmt - fm->per_cpu);
+  vlib_zero_simple_counter (&gtm->upf_simple_counters[UPF_FLOW_COUNTER],
+			    fmt - fm->per_cpu);
+  upf_counter_unlock (gtm);
+
```
Comment (RoadRunnr) [inline upf/flowtable_init.c:93]: putting the counters in flowtable_main_t might make more sense. simple_counters are per CPU thread, repeating that by using the index and CPU id does not make sense.

Or is there a specific reason you want to read the per CPU counters?

File: upf/upf.c [hunk 2]
```diff
@@ -359,6 +359,29 @@ upf_init (vlib_main_t * vm)
 					FIB_SOURCE_PRIORITY_HI,
 					FIB_SOURCE_BH_SIMPLE);
 
+  clib_spinlock_init (&sm->upf_counter_lock);
```
Comment (RoadRunnr) [inline upf/upf.c:362]: the look is not really needed.
Simple counters are per CPU anyway and if you don't use the index as per CPU id, then no locking is required.

**Review by RoadRunnr — CHANGES_REQUESTED** [review]

File: upf/flowtable_init.c (hunk 1 above)
Comment (sergeymatov) [reply upf/flowtable_init.c:93]: I've just following your suggestion to make it per-cpu, but I suppose it was only suitable for gauges.
Keeping in mind that counters are per-cpu vectored more logical is to have in in flowtable_main for sure.
Will be changed.

File: upf/upf.c (hunk 2 above)
Comment (sergeymatov) [reply upf/upf.c:362]: sure. will be fixed

File: upf/flowtable.c [hunk 3]
```diff
@@ -168,12 +168,17 @@ expire_single_flow (flowtable_main_t * fm, flowtable_main_per_cpu_t * fmt,
...
+      upf_main_t *gtm = &upf_main;
       upf_debug ("Flow Remove %p", f);
       pool_put (fmt->timers, e);
 
       /* hashtable unlink */
       flowtable_entry_remove (fmt, f);
 
+      vlib_decrement_simple_counter (&gtm->upf_simple_counters
+				     [UPF_FLOW_COUNTER],
+				     0, fmt - fm->per_cpu, 1);
```
Comment (RoadRunnr) [inline upf/flowtable.c:180]: this and the other places where the FLOW_COUNTER is used are wrong. You mixed up the `thread_index` with the `index`

```suggestion
      vlib_decrement_simple_counter (&gtm->upf_simple_counters
				     [UPF_FLOW_COUNTER],
				     fmt - fm->per_cpu, 0, 1);
```

@ivan4th what do think about the way that is used to calculate the `thread_index`? `fmt - fm->per_cpu` is correct, but also quite strange.
I would rather prefer to use `vlib_get_thread_index ()` instead.

Comment (ivan4th) [reply upf/flowtable.c]: I agree that `vlib_get_thread_index ()` is better there

**Comment by ivan4th** [comment]
@RoadRunnr all objections fixed as expected?

**Review by RoadRunnr — APPROVED** [review]
LGTM

**Pull request merged**